config/clusters/callysto/common.values.yaml

```
@@ -65,67 +65,68 @@ jupyterhub:
-            default=None,
-            # This must always be set!
-            allow_none=False,
-            help="""
-            Allow access to users with verified emails belonging to these domains.
+            """
+            Custom override of CILogonOAuthenticator to allow access control via
+            one property (email) while the username itself is a different
+            property (oidc), and to allow `allowed_domains` to involve wildcards
+            like *.
+
+            This allows us to restrict access based on the *email* of the
+            authenticated user without having to store the actual email in our
+            servers or use it as a username.
+            """
 
-            Allows glob patterns but not regexes. So you can allow access to
-            *.edu, for example. See https://docs.python.org/3/library/fnmatch.html for
-            list of allowed patterns.
-            """,
-            config=True
-          )
+            async def check_allowed(self, username, auth_model):
+                """
+                Replaces the default implementation of check_allowed as it
+                otherwise would error, not finding a `@` symbol in the non-email
+                username.
 
-          async def authenticate(self, *args, **kwargs):
-            authenticated_user = await super().authenticate(*args, **kwargs)
+                Mostly a copy of OAuthenticator 16.1, for comparison see:
+                https://github.com/jupyterhub/oauthenticator/blob/16.1.0/oauthenticator/cilogon.py#L349-L373
+                """
+                if await super(CILogonOAuthenticator, self).check_allowed(username, auth_model):
+                    return True
 
-            email = authenticated_user['auth_state']['cilogon_user']['email']
+                user_info = auth_model["auth_state"][self.user_auth_state_key]
+                user_idp = user_info["idp"]
 
-            domain = email.rsplit('@', 1)[1]
-            for ad in self.allowed_domains:
-              # fnmatch will allow us to use wildcards like * and ?, but not the full
-              # regex. For simple domain matching this is good enough. If we were to
-              # use regexes instead, people will have to escape all their '.'s, and since
-              # that is actually going to match 'any character' it is a possible security hole.
-              if fnmatch(domain, ad):
-                return authenticated_user
+                idp_allow_all = self.allowed_idps[user_idp].get("allow_all")
+                if idp_allow_all:
+                    return True
+
+                user_info = auth_model["auth_state"][self.user_auth_state_key]
+                user_idp = user_info["idp"]
+
+                idp_allowed_domains = self.allowed_idps[user_idp].get("allowed_domains")
+                if idp_allowed_domains:
+                    # this was the part we wanted to replace
+                    email = auth_model["auth_state"]["cilogon_user"]["email"]
```

Just to be consistent and to make sure this doesn't change [upstream](https://github.com/jupyterhub/oauthenticator/blob/2fa1c4818f7ba3444f3e2e44f8ddb38127553ffd/oauthenticator/cilogon.py#L45) and forget about it here:

```suggestion
                    email = auth_model["auth_state"][self.user_auth_state_key]["email"]
```